src/Pact/Compile.hs

```diff
@@ -107,49 +108,66 @@ mkHash msg h el = case fromText' h of
       Left e -> mkInfo el >>= \i -> syntaxError i $ msg ++ ": bad hash: " ++ e
       Right mh -> return mh
 
-justDocs :: Text -> Maybe Meta
-justDocs docs = Just $ Meta docs def
+justDocs :: Text -> DocModel
+justDocs docs = DocModel (Just docs) M.empty
 
-mkMeta :: Text -> [Exp] -> Compile (Maybe Meta)
-mkMeta docs ps = Just . Meta docs . M.fromList <$> mapM toMetas ps
+mkModel :: Exp -> Compile (M.Map Text Exp)
+mkModel ps = uncurry M.singleton <$> toMetas ps
   where toMetas (EList' [EAtom' k,v]) = return (k,v)
-        toMetas bad = mkInfo bad >>= \i -> syntaxError i "mkMeta: malformed meta form, expected (atom expression)"
+        toMetas bad = mkInfo bad >>= \i -> syntaxError i "mkModel: malformed model, expected (atom expression) where atom is one of {property, properties, invariant, invariants}"
 
 -- | A meta-annotation, which includes either form:
 --
 -- * `"docstring"`
--- * `(meta ...)`
-pattern MetaExp :: Compile (Maybe Meta) -> Exp
-pattern MetaExp meta <- (doMeta -> Just meta)
-
-doMeta :: Exp -> Maybe (Compile (Maybe Meta))
-doMeta exp = case exp of
-  EList' (EAtom' "meta" : ELitString docs : ps) -> Just (mkMeta docs ps)
-  ELitString docs                               -> Just (pure (justDocs docs))
-  _                                             -> Nothing
-
--- | A body with a possible meta-annotation
-pattern MetaBodyExp :: Compile (Maybe Meta) -> [Exp] -> [Exp]
+-- * `@doc ...` / `@meta ...`
```

Comment:
Hmm ... the approach here "gets the job done" but it's nasty from an extensibility perspective. Compile's approach -- pattern matching on lists to recognize code -- is pretty much broken here.

Can we add a note something to the effect of "TODO: make meta tag recognition extensible via proper token parsing"? It's clear an upcoming fix will be to move Compile to a proper parser over Exp, especially as we confront the parser overhaul in #137 .

Reply:
I started with a more extensible approach but decided that this was simpler given that there are just a handful of cases and as far as I know we don't have plans to allow anything other than `@doc` and `@model` for now.

Reply:
note added